### ISSUE TITLE
Define model geometry extend in or to the world builder.

### ISSUE THREAD

**Issue by MFraters** [GeodynamicWorldBuilder/WorldBuilder#331]
This came up in a discussion with @bangerth during the aspect biweekly meeting and is related to https://github.com/GeodynamicWorldBuilder/WorldBuilder/issues/318, and may be of interest to @alarshi. The post has become a little bit longer than I hoped, but I think it is a good summary of the issue and possible solutions.

The main issue is that it is currently very hard to optimize for especially spherical domains, because the world builder is designed to not know anything about the domain, but fully rely on the depth and height information provided at each request. The advantage of this approach is that there is no room for error by the user when building a model, because the code calling the world builder defines the exact world in the call. For example in ASPECT the user defines the information about the world in the ASPECT input file and when aspect requests the temperature at a certain height and depth, the radius of the earth in a spherical geometry is automatically defined based on that information.

This prevents mistakes and provides a lot of flexibility, because you can change your aspect model domain without having to worry about the world builder file and even change the size of the world by modifying the the values you provide to the get function. The downside is that the GWB only knows the size of the planet at the time it is requested. This also means that many optimizations, especially related to the slabs (which are expensive to compute) are simply not possible. 

A different issue, which was mentioned by @tjhei, is that this also means that a GWB file does not uniquely define a model. It only defines a unique model in combination with for example an ASPECT input file or a world builder visualizer input file. This may not be what users expect.

Here are some of the proposed solutions:

1. Provide the radius of the earth or model box depth/height for a spherical or Cartesian geometries respectively as a **special optimization input parameter** in the GWB input file. This would be be easy to implement, but it would hand of the responsibility to the user, creating unexpected results for users who are just reusing files they got from others. 
2. Provide the radius of the earth or model box depth/height for a spherical or Cartesian geometries respectively as a **normal input parameter** in the GWB input file in the coordinate system section. This would be be easy to implement and would change the world builder file into a complete description of the initial conditions, but it would allow for incompatibilities between the GWB and ASPECT input files. In the current structure that would mean adding an assertThrow (release mode) in every query to prevent this. I think this would be error prone and annoying to users (see option 5). 
3.  Provide the radius of the earth or model box depth/height for a spherical or Cartesian geometries respectively as **a parameter of the construction of the world for optimization**. This would allow for the optimizations and would only rely on the correct information be send by for example ASPECT (which is already the case), and can be tested in debug mode. This would be relatively easy to implement. It probably be implemented in a backwards compatible way, but I am not against breaking backwards compatibility if that leads to cleaner code (the GWB is still in beta). 
4. Provide the radius of the earth or model box depth/height for a spherical or Cartesian geometries respectively as **a parameter of the construction of the world and remove the request for the depth in the get functions**. Having the radius of the earth means that you don't need both the depth and height of a model anymore, just one of them (height is already part of the coordinate of the point requested). This would allow for the optimizations and would only rely on the correct information be send by for example ASPECT (which is already the case). This would be require a new get interface. I think the new interface could be additional, with the old interface just leading to the new one and ignoring one of the parameters. The old interface could then be deprecated. The changes themself would not be very hard to implement, but there will be a lot of places where changed need to be made. An other option would be to just make the change in an incompatible way. This would result in cleaner code but would require more changes in the short term. As I noted in point 3, I am not against breaking backwards compatibility if that leads to cleaner code (the GWB is still in beta). 
5. Provide the radius of the earth or model box depth/height for a spherical or Cartesian geometries respectively as a **normal input parameter** in the GWB input file in the coordinate system section **and as a parameter of the construction of the world**. This way it is possible to make the optimizations, check for inconsistencies between GWB and ASPECT at the world creation, and let a GWB file uniquely define the initial conditions. A good default could be set for spherical models, but that is kind of impossible for Cartesian models. I notice that I often change the depth of my Cartesian models, and having to change it in two different files would slow down progress. 
6. Provide the radius of the earth or model box depth/height for a spherical or Cartesian geometries respectively as a **normal input parameter** in the GWB input file in the coordinate system section **, as a parameter of the construction of the world and replace the request for either depth or height in the get functions**. This has more or less the same advantages and disadvantages as point 5, but since some information has become unneeded, it could be removed. Although it may be computationally more efficient though to just let the external program (e.g. ASPECT) provide this information so that is doesn't have to be recomputed every time. 

My favorites are currently point 3 and 5. I like that in 5 the GWB completely and uniquely describes the initial conditions, but I don't really like if I change the depth of the Cartesian model (which is reasonably common for me), I will need to change two numbers in two different files. That may of course be fine if the advantage is large enough. 

Since point 3 is a prerequisite for 5, if we agree that either 3 or 5 are the way to go, we can already implement 3. Before doing that I will keep this issue open for a few days to give everyone who wants to read and respond to this a chance. My opinion of this is not set in stone and I will be happy to be convinced otherwise or show options I didn't think of!

**Comment by tjhei** [comment]
> 5\. A good default could be set for spherical models, but that is kind of impossible for Cartesian models. I notice that I often change the depth of my Cartesian models, and having to change it in two different files would slow down progress.

For spherical models you would define a radius (distance of surface to center of earth), while the height of the simulation (maximum depth) is independent of that. I would argue that for Cartesian models you would not specify the depth of the simulation but also the distance to the center of the planet. Maybe this is infinity for a box? This way, you can change the maximum depth of a Cartesian model in the same way as you would for a spherical (or spherical chunk).

**Comment by MFraters** [comment]
Thanks for the input Timo! I had to think a bit more about it because I now realize that I did not split in my mind two things which can be separated. One is whether you can use the world builder file to completely define the initial conditions. This definition is only in a depth coordinate system. The other one is whether the world builder knows how to convert a depth based coordinate system to a coordinate system where the origin is not at the surface. 

1. Currently the world builder knows both by asking  both the height and depth of the requested point. This defines the conversion from the depth coordinate to the height coordinate (an vice versa).
2. Providing the radius for spherical and an "infinite radius" for Cartesian means that the world builder would not require to ask for the depth anymore in the spherical case, but it would still need it in the Cartesian case. So the location of the origin with respect to the surface is known for the spherical model, but not for the Cartesian case, but I would argue that the model is fully defined in both cases.
3. Providing the radius for the spherical case and the model depth (basically the location of the origin with respect to the surface), would mean that the depth doesn't need to be requested anymore, since that can be calculated. In this case both the initial condition and the conversion between the coordinate systems are fully defined for both spherical and Cartesian coordinate systems when the world is created, in stead of when the properties (like temperature) are requested.

For point 2 (and point 3 I guess), the height is actually not needed and the world builder could instead just request a point of (x,y,depth) for cartesian and (long,lat,depth) for spherical. Then no conversion is needed anymore.

**Comment by bangerth** [comment]
@MFraters My apologies for sitting on this for so long.

Here is my suggestion, which may be what you already implemented in #345 -- I will see about that next -- but let me spell it out because it helps me understand things myself:

* WorldBuilder is independent of any external projects, and so it would be nice if it had a complete view of how the world looks independently of any of the other projects that use it. As such, I think the right place to specify a radius would be in the input file. Then, one could create a WorldBuilder model from only the input file and external codes never have to set anything, they just query a blackbox. All optimizations can be done at the time of reading the input file.
* ASPECT of course has its own notion of the world, specified in its own input file. Since ASPECT uses WorldBuilder, but not the other way around, the responsibility for checking whether the two worlds are the same would rest with ASPECT. My take is that it should come in the form of an assertion in the ASPECT code right after initializing both of the geometries that looks like
```
  AssertThrow (world->get_max_depth() == geometry_model->get_max_depth(), ExcMessage ("Don't do that!"));
```
or whatever the appropriate calls for the max depth are.

I think this would provide WorldBuilder with the cleanest path forward and makes clear who is responsible for what. In particular, WorldBuilder is responsible for defining a black-box model that is complete and can be queried by external codes. The fact that these external codes might have their own notion of how the world looks like shouldn't be WorldBuilder's responsibility. In particular, if you require external codes to pass a radius in at construction time, then you presume that external codes in fact have a separate notion of the model geometry -- but this may not in fact be true, it is easy to imagine someone building a code that simply gets *all* of its geometric information from WorldBuilder and, consequently, would not know what actually to pass to the constructor of WorldBuilder regarding a radius or depth.

**Comment by MFraters** [comment]
Thanks for you input Wolfgang, it is really appreciated! I had to think a bit about it, because what you are proposing is quite a shift in responsibility. I am writing down my thoughts on this below, also partly to just write them down in a structured way for myself. Further comments and though would be really appreciated. 

The current approach provides for an extreme amount of flexibility where the user of a code like ASPECT only has to set the dimensions in ASPECT once, and never worry about it anywhere else. With your and Timo's input I have started to realize that the flip side of this approach is not only that is becomes very hard to precompute things, but also that the world builder file not fully defining the world is very counter intuitive. I can clearly see the appeal of having a world builder file which fully defines a world.

The appeal of the current approach is that it eliminates whole classes of mistakes in the usage of the world builder because as long as you provide the point and the depth of the point correctly, there is no room for error. You define your model once in the calling program, with computes the depth of that point, and it will automatically work. Making this change does adds complexity and creates the room for error, for example between what is defined in the calling program and what is defined in the world builder. I think the upsides of changing the current approach outweigh the upsides of the current approach. So I am in favor of defining the depth of the model in the world builder file. 

You do make a very good argument for not adding the maximum depth to the constructor. I had not thought of a case like that, and that is clearly something which needs to be supported by the world builder. So this does rule out the current proposed solution in pull request #345. This will may also mean that a new solution would most likely not cause an api breakage (see the last paragraph for a different potential breakage), which I think is generally a good thing. 

This does not mean we are not able anymore to check for consistency between the maximum depth in the world builder file and the one (implicitly or explicitly) defined in the external program. Because in every call we still get both the position (height) and the depth of that position, we can still reconstruct the maximum depth, like we do now. Adding a debug mode WBAssert would be an easy way for external programs to check whether their implementation is correct. We could also check it in release mode to make very sure that everything is correct, but at that point it is an implementation issue by the calling program which should have been caught in debug mode. The overhead of that check should not be big though, so it may be nice to get some extra security for very little overhead. 

Whether an assert would be desired or needed in the end this comes back to your point about who is responsible and if it in some way inhibits (unexpected) valid use. I do agree that it would make sense that since ASPECT is using the world builder, the responsibility for the correct usage of the world builder lies with ASPECT. I know that I may be a bit overprotective, but that is in part because I know that I will be implementing and using it in ASPECT, and I am very aware how many mistakes I make ;)

If we would decide that the checking would not be needed, the interface could be simplified. The depth would no longer be needed. Because I currently think having at least an assert (maybe even an assertthrow) would have value, I don't think we should go this route, but I did want to point it out.

**Comment by bangerth** [comment]
I needed to write out my thoughts as well -- that sometimes really helps clarify thoughts.

I'm glad we're agreed about the general idea that WorldBuilder keeps a complete view of the world and this way also becomes usable by other codes that do not have a separate representation outside WorldBuilder. If I understand your question about the assert correctly, what you are wondering if whether there is value in passing *redundant* information into functions (i.e., information that can either be computed from the other function arguments, or from data stored elsewhere in a WorldBuilder model) so that one can write additional asserts. That would be equivalent to writing `matrix.get_element(i,m, j,n)` instead of `matrix.get_element(i,j)` even though the `matrix` object knows its size `(m,n)`. I would generally advise against this kind of thing because it incurs overhead every time you call the function; if a calling function wants to convince itself that it knows the matrix dimensions, it can always do so with an extra `assert (m == matrix.n_rows())`, but it has to do that only in one place and not, for example, in the inner loop where it calls `matrix.get_element(...)`. In your case, if we want to convince ourselves that WorldBuilder's idea of maximal depth agrees with ASPECT's idea of maximal depth, we should do so once both codes have read their input file. There is no need to do that every time we want to evaluate some material property.

**Comment by MFraters** [comment]
hmm, I now see that removing the depth from the external query functions indeed follows logically from fully defining the world in the world builder. 

This would mean a major api breakage, although it would be easy to deprecate the old interface and just forward it to the new one and then remove it in version 0.6.

This does give an opportunity to rethink the interface a bit more. Currently these are the (3d) interfaces:
```
  World::temperature(const std::array<double,3> &point_,
                     const double depth,
                     const double gravity_norm) const
```
```
  World::composition(const std::array<double,3> &point_,
                     const double depth,
                     const unsigned int composition_number) const
```
```
  World::grains(const std::array<double,3> &point_,
                const double depth,
                const unsigned int composition_number,
                size_t number_of_grains) const
```

Depth will be removed from all of them, but I never really liked the the gravity norm in the temperature interface. Now that the the world builder is moving to be more self contained, I like it even less. The reason that it is needed is that some temperature models need the gravity norm (anything including an adiabatic gradient for example). Since the gravity can change spatially, a single number would not be enough. So it would mean adding a new plugin system for gravity. 

Adding this system could lead to more work for the user in ASPECT to check that the gravity models are the same, but to help with the gravity could be provided as a output which could be used by an ASPECT gravity plugin. This would allow the temperature interface to become: 
```
  World::temperature(const std::array<double,3> &point_) const
```
It would also mean that gravity information would become available to all parts of the world builder, which may be useful for other things as well (no idea what yet). What do you think about this?

**Comment by bangerth** [comment]
Ah, that's an interesting question too from the design perspective. In some sense, if you want your WB model to be self-contained, then the model itself needs to have everything necessary to compute the temperature. That is, you probably want to make sure that it has a gravity model built in if the gravity determines the temperature. You can think of this as simply a parameter: The temperature increases with depth like (I'm making this up) `T0 * exp(alpha*z)` where `z`=depth. Here, `alpha` will then simply be a growth rate and, in your case, will somehow depend on gravity. So your input file needs to either contain a place where someone can specify what `alpha` is, or if `alpha` is computed from other quantities like `alpha=alpha(g,h,j)` then what `g,h,j` are.

These are parameters that could be specified globally for the entire model, or for each plate/slab/... depending on the situation you are trying to model. But in the end, like for the radius, the right approach would likely be to have these be parameters of WB, not of some external program.

There are of course situations where you might want to be more general. For example, one could imagine that you wanted the temperature to depend on some chemical composition, and that might change with time. In that case, WB input files can not capture the complicated dynamics of `g(t),h(t),j(t)` and it *has* to be an argument passed to that function -- though in that case one might wonder how general you want that interface to be, because surely there are then situations where `g,h,j` are no longer enough and the temperature now also depends on `k,l,m` and you would have to extend the list of arguments. 

I don't know how general you want to be, and that determines what arguments the function should take. In similar circumstances in ASPECT -- say, in the material models -- we've simply abstracted this and put everything into `AdditionalMaterialInputs/Outputs` that each material model interprets in its own way. These objects then encapsulate whatever data needs to be passed, without having to spell out what they are in the interface descriptions.

**Comment by MFraters** [comment]
I don't think I am currently interested in adding time to the world builder, although it is an interesting thought ;)

Yes, I agree that going this route will require some kind of gravity model. I see two options. 
1. Have each model which needs gravity, compute it based on depth internally. This could have an option to have a global parameter which describes the methods. I think that that is what you are describing.
2. Make a full gravity model plugin system. The temperature plugins can then just call the gravity plugin system to find out what the gravity is. The advantage of this is that it could also provide aspect with the gravity through a plugin, making sure that aspect and the world builder remain consistent with each other. The gravity models which are currently in aspect do not depend on any other parameters and most of them (except ascii data and function) are either constant or linear. Those would be very easy to implement. The only concern with this approach I have is performance, but I don't think it needs to be. Looking at the plugins in aspect, you don't need to know what feature you are in, so the function can be just as cheap as the counterparts in aspect.

As for the interface, the issue here is that I need to have stable wrappers functions for C (and other languages) at some point, which means that I won't be able to change them until the next major release (after the 1.0.0 release). That also means that adding an additional inputs structure is going to have the same effect as just adding the values to the interface function. Any change in the structure breaks compatibility (at least I think, please correct me if I am wrong). So going towards a more full description of the world which is being build is probably the way to go when these issues are encountered. So in your example that would be describing the chemical composition properties internally, which is currently already happening for things like the heat capacity and thermal expansion.

**Comment by bangerth** [comment]
I have no particular opinion about which approach to use, primarily because I don't quite know where you use gravity. But does it matter? ASPECT has a whole plugin system for gravity because we want to be able to model non-vertical gravity where necessary. That might be useful for irregularly shaped celestial bodies, for example, or because it affects the direction of buoyancy. But I suspect that for what you want to do (assuming that all you're interested in is the adiabatic state), using a vertical gravity model is entirely sufficient, and that in that case gravity is a *global* property of your model, not one that you need to describe for each plate/slab/feature separately. So a slab could ask the global WB object for what gravity is, and use that, and the global WB object would get that out of a parameter that's set in the input file.

**Comment by MFraters** [comment]
Now that #401 (which greatly simplified some complex parts of the world builder) is merged, I have started to think again about about this issue. I want to get rid of the array, so that it is more clear what the input means. I have considered the the following options for the new interface (showing 3d temperature here):

1. `World::temperature(const double x, const double y, const double z, const double depth)` This is basically what is currently implemented. **This option is possible**, but there is redundant information in here which could be in conflict. 
2. `World::temperature(const double x, const double y, const double depth) const` This works great for cartesian, but requires to compute `z` in spherical, which can be done up to the sign (can't know what side of the world the point is). **So this option is not possible.**
3. `World::temperature(const double x, const double y, const double z) const` This option is possible, but requires an actual max depth for cartesian models (in stead of the infinity which is now used). **This option is possible, but not great**, since it requires all cartesian models to specify the max depth of the model which in my experience can change a lot when testing.
4. `World::temperature(const double x_or_long, const double y_or_lat, const double depth) const` This works great for Cartesian, and spherical, but requires different input for cartesian and spherical model. **This option is possible, but requires spherical coordinates as input for spherical model.** Although this actually kinda sound logical, it would mean an extra step/extra code in codes like aspect which work fully in Cartesian coordinates. 

In case 4 the interface would probably look like this: 
`World::temperature(const double x_or_opening_angle, const double depth) const`

My current ranking of the options is 4,1,3, but I am happy to change my mind.

**Comment by tjhei** [comment]
Have you considered providing a set of several functions to query information in different ways (cartesian, lat/long, etc.)?

It would be silly in ASPECT to convert from xyz->lat/long just so that you will convert back immediately, for example. Other codes might prefer that, though.

**Comment by MFraters** [comment]
Thanks for your comment! 

> Have you considered providing a set of several functions to query information in different ways (cartesian, lat/long, etc.)?

In my mind that sounded like an adding extra maintenance burden, but if I think back of how much maintenance time the current interface has taken me, it is basically zero. So that would be an option. 

> It would be silly in ASPECT to convert from xyz->lat/long just so that you will convert back immediately, for example. Other codes might prefer that, though.

Yes, good point. From the perspective of the world builder it doesn't really matter because it needs both in the end, but I can definitely see that it would make use from aspect more difficult. 

The more I think about it, the less value I see of changing the current interface. The only advantage it that it is removing some redundant information, but the disadvantage is that the alternatives are either worse or more complicated (splitting up cartesian and spherical models for example). The only change which really needs to be made is the removal if the gravity parameter from the temperature interface, but the rest seems more trouble than it is worth...

**Comment by MFraters** [comment]
I think this issue is now resolved with the merger of #404.
